elfo-network/src/frame/read.rs

```diff
@@ -0,0 +1,224 @@
+    unsafe { &mut *(slice as *mut [MaybeUninit<u8>] as *mut [u8]) }
+}
+
+impl FramedReadStrategy for LZ4FramedRead {
+    fn read(&mut self) -> Result<FramedReadState<'_>> {
+        let (compressed_size, position) = match self.state {
+            LZ4DecodingState::FrameDecompression => {
+                let compressed_slice = unsafe {
+                    slice_assume_init_ref(&self.compressed_buffer[self.position..self.filled])
+                };
+                let lz4_state = self
+                    .decompressed_buffer
+                    .decompress_frame(compressed_slice)?;
+                match lz4_state {
+                    DecodeState::NeedMoreData { length_estimate } => {
+                        // Decoder requested for more data to be read. We round up the number of
+                        // requested bytes to be a multiple of the chunk size to avoid
+                        // reading too few bytes per syscall.
+                        debug_assert!(length_estimate > self.filled - self.position);
+                        let rounded_length = ((length_estimate + BUFFER_CHUNK_SIZE - 1)
+                            / BUFFER_CHUNK_SIZE)
+                            * BUFFER_CHUNK_SIZE;
+
+                        let postfix_free_space = self.compressed_buffer.len() - self.filled;
+                        if rounded_length <= postfix_free_space {
+                            // There is enough space in the buffer for the
+                            // requested data to be read
+                            // into.
+                        } else {
+                            // Shift the data into the beginning of the buffer.
+                            unsafe {
+                                let buffer_start = self.compressed_buffer.as_mut_ptr();
+                                let data_start = buffer_start.add(self.position);
+                                let len = self.filled - self.position;
+                                std::ptr::copy(data_start, buffer_start, len);
+
+                                self.position = 0;
+                                self.filled = len;
+                            }
+
+                            let total_free_space = self.compressed_buffer.len() - self.filled;
+                            if rounded_length > total_free_space {
+                                // If there is still not enough space after the shift, we need to
+                                // reallocate.
+                                let new_buffer_size = self.filled + rounded_length;
+                                self.compressed_buffer
+                                    .resize(new_buffer_size, MaybeUninit::uninit());
+                            }
+                        }
+
+                        let compressed_slice = unsafe {
+                            slice_assume_init_mut(&mut self.compressed_buffer[self.filled..])
```

Comment:
I am pretty sure this is UB. We are guaranteed to have uninit data here. I would return `&mut [MaybeUninit<u8>]` here, but the `OwnedReadHalf::read` accepts only `&mut [u8]`